src/main/java/org/jnode/fs/xfs/directory/BlockDirectoryEntry.java

```diff
@@ -46,63 +84,25 @@ public class BlockDirectoryEntry extends XfsObject {
     public BlockDirectoryEntry(byte[] data, long offset, boolean v5) {
         super(data, (int) offset);
         this.isV5 = v5;
-        isFreeTag = getUInt16(0) == 0xFFFF;
-        if (!isFreeTag()) {
-            nameSize = getUInt8(8);
-            iNodeNumber = getInt64(0);
-            byte[] buffer = new byte[nameSize];
-            System.arraycopy(data, (int) offset + 9, buffer, 0, nameSize);
+        freeTag = getUInt16(0) == 0xFFFF;
+
+        if (!freeTag) { // it is xfs_dir2_data_entry
+            iNodeNumber = readInt64();
+            nameLength = readUInt8();
+            byte[] buffer = new byte[nameLength];
+            System.arraycopy(data, (int) offset + 9, buffer, 0, nameLength);
             name = FSUtils.toNormalizedString(buffer);
-        } else {
-            nameSize = 0;
+            unusedLength = 0;
+            skipBytes(nameLength);
+        } else { // it is xfs_dir2_data_unused
+            nameLength = 0;
             iNodeNumber = 0;
             name = "";
-        }
-    }
-
-    /**
-     * Gets the inode number of this entry.
-     *
-     * @return the inode number
-     */
-    public long getINodeNumber() {
-        return iNodeNumber;
-    }
-
-    /**
-     * Gets the name size of this entry.
-     *
-     * @return the name size.
-     */
-    public int getNameSize() {
-        return nameSize;
-    }
-
-    /**
-     * Gets the name of this entry.
-     *
-     * @return the name entry.
-     */
-    public String getName() {
-        return name;
-    }
 
-    /**
-     * Gets the offset of the directory block.
-     *
-     * @return the offset directory block
-     */
-    public long getOffsetFromBlock() {
-        return getUInt16(getNameSize() + 9);
-    }
+            unusedLength = readUInt16();
```

Comment:
I think this needs to skip the size of the `freeTag` read - the `freeTag` check above doesn't increment `XfsObject.offset`.

Reply:
yeah, and the unit test didn't cover it and the header thing, I'll change it here and add some test as well.